### PR TITLE
[E-103104] - Port gaze_estimation_demo gapi demo to OV API2.0

### PR DESCRIPTION
**Pull request by DariaMityagina**
### Tiket:
103104

### Description:
G-API Background Subtraction Demo, Classification Benchmark C++ G-API Demo, G-API Gaze Estimation Demo and other - fail on NPU with the following error: `dynamic shape issues`
```
get_shape was called on a descriptor::Tensor with dynamic shape
```

- OpenVINO is actively removing API 1.0 from their code
- Both G-API framework and G-API demos use API 1.0 --> which leads to disruption of certain functionality
Conclusion: We need to port the G-API demos to API2.0 to make them work.

- [x] gaze_estimation_demo
OK: Fixed on NPU. Works with a **single image** and with a **video input**.
FAIL: But still has issues in case **folder with images** is provided (in CI) - WIP
` Failed to copy data from cv::Mat to ov::Tensor. Data type or number of elements mismatch. ov::Tensor: u8 [1,490,1183,3] and cv::Mat: 8UC3 1180x490`
It seems to fail if the input folder contains images of different sizes. It works fine with the first image from the folder and does not work with the second (if it is a different size).

### CODE REVIEW

File: demos/tests/cases.py [hunk 1]
```diff
@@ -255,7 +255,7 @@ def single_option_cases(key, *args):
             test_cases=combine_cases(
         TestCase(options={'-no_show': None,
             **MONITORS,
-            '-i': TestDataArg('coco128/images/train2017/')}),
+            '-i': DataPatternArg('coco128-every-480x640x3')}),
```
Comment (DariaMityagina) [inline demos/tests/cases.py:258]: This case is closer to the video input (since frame size doesn't change)

Comment (DariaMityagina) [reply demos/tests/cases.py:258]: For some reason there are issues with reading video input in github actions: https://github.com/openvinotoolkit/open_model_zoo/actions/runs/7408585487/job/20157102210?pr=3887

image

Comment (Wovchena) [reply demos/tests/cases.py:258]: You may check OpenCV cmake flags. I heavily reduced build time by disabling modules

Comment (DariaMityagina) [reply demos/tests/cases.py:258]: I see, will check them. Thanks!

File: demos/tests/run_tests.py [hunk 2]
```diff
@@ -51,12 +51,15 @@
...
+import shutil
+import requests
+
 scopes = {
     'base': importlib.import_module('cases').DEMOS,
     'performance': importlib.import_module('performance_cases').DEMOS,
 }
 COCO128_URL = "https://ultralytics.com/assets/coco128.zip"
-
+VIDEO_URL = "https://storage.openvinotoolkit.org/data/test_data/videos/head-pose-face-detection-male.mp4"
```
Comment (DariaMityagina) [inline demos/tests/run_tests.py:62]: This is the same video used in the `gaze_estimation_demo` readme file ([here](https://github.com/openvinotoolkit/open_model_zoo/tree/master/demos/gaze_estimation_demo/cpp_gapi))
image

image

File: .github/workflows/cpp_gapi-demos.yml [hunk 3]
```diff
@@ -35,10 +35,12 @@ jobs:
       - if: ${{ !steps.cache.outputs.cache-hit }}
         name: Compile OpenCV
         run: |
+          sudo apt install pkg-config
+          sudo apt install ffmpeg libavcodec-dev libavformat-dev libavutil-dev libswscale-dev libavresample-dev
```
Comment (DariaMityagina) [inline .github/workflows/cpp_gapi-demos.yml:39]: Required for `-DWITH_FFMPEG=y`.

File: demos/tests/cases.py [hunk 4]
```diff
@@ -266,6 +265,10 @@ def single_option_cases(key, *args):
             '-m_fd',
             ModelArg('face-detection-adas-0001'),
             ModelArg('face-detection-retail-0004')),
+        single_option_cases(
+            '-i',
+            str('/tmp/video.mp4'),
+            DataPatternArg('coco128-every-480x640x3')),
```
Comment (DariaMityagina) [inline demos/tests/cases.py:271]: Both options:
- video input
- folder with images as input

File: .github/workflows/cpp_gapi-demos.yml (hunk 3 above)
Comment (DariaMityagina) [reply .github/workflows/cpp_gapi-demos.yml:39]: This does not affect OpenCV build time.

**Comment by DariaMityagina** [comment]
@Wovchena hi! Could you please review this PR?

File: demos/tests/cases.py (hunk 1 above)
Comment (DariaMityagina) [reply demos/tests/cases.py]: Done, thanks!

**Review by Wovchena — CHANGES_REQUESTED** [review]

File: .github/workflows/cpp_gapi-demos.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (DariaMityagina) [reply .github/workflows/cpp_gapi-demos.yml]: Done, thanks! https://github.com/openvinotoolkit/open_model_zoo/pull/3891/commits/1ab0f38e5a064c649c2460f63373b92eebae4386

File: demos/tests/run_tests.py [hunk 5]
```diff
@@ -247,6 +249,14 @@ def main():
         with ZipFile(BytesIO(zipresp.read())) as zfile:
             zfile.extractall(args.test_data_dir)
 
+    try:
+        r = requests.get(VIDEO_URL)
+        with open("video.mp4", 'wb') as f:
+            f.write(r.content)
+            print('Video for demos downloaded successfully')
+    except requests.exceptions.HTTPError as err:
+        print(err)
```
[the remark this answers is not in the document]
Comment (DariaMityagina) [reply demos/tests/run_tests.py]: Done, thanks! https://github.com/openvinotoolkit/open_model_zoo/pull/3891/commits/1ab0f38e5a064c649c2460f63373b92eebae4386

**Review by Wovchena — APPROVED** [review]

**Comment by dmatveev** [comment]
Thanks a lot @Wovchena @DariaMityagina !

**Pull request merged**